File: solutions/Strings/0131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
     List<List<String>> result=new ArrayList<>();
     List<String> path =new ArrayList<>();
     part(s,0,path,result);  
     return result; 
    }
    private void part(String s,int index,List<String> path,List<List<String>> result)
    {
        if(index==s.length())
        {
            result.add(new ArrayList(path));
            return;
        }
        for(int end=index;end<s.length();end++)
        {
            String substring=s.substring(index,end+1);
            if(isPalindrome(substring))
            {
                path.add(substring);
                part(s,end+1,path,result);
                path.remove(path.size()-1);
            }
        }
    }
    private boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) return false;
        }
        return true;
    }
}

File: solutions/Strings/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:

        result=[]

        def is_palindrome(sub):
            return sub == sub[::-1]
        
        def backtrack(start,path):
            if start==len(s):
                result.append(path[:])
                return
            #try every cut

            for end in range(start,len(s)):
                substring=s[start:end+1]

                #check if palindrome and continue only if palindrome
                if is_palindrome(substring):
                    path.append(substring)

                    #explore
                    backtrack(end+1,path)

                    #undo
                    path.pop()
        backtrack(0,[])
        return result
